ambry-network/src/main/java/com/github/ambry/network/NettyServerRequestResponseChannel.java

```diff
@@ -80,30 +104,91 @@ public void closeConnection(NetworkRequest originalRequest) throws InterruptedEx
     }
   }
 
-  /** Get the next request or block until there is one */
+  /** Get the next request or block until there is one.
+   * @return*/
+  @Override
+  public NetworkRequestBundle receiveRequest() throws InterruptedException {
+
+    NetworkRequestBundle networkRequestBundle;
+    NetworkRequest requestToServe;
+    Collection<NetworkRequest> requestsToDrop;
+    do {
+      networkRequestBundle = networkRequestQueue.take();
+      NetworkRequest unqueuedRequest;
+      while ((unqueuedRequest = unqueuedRequests.poll()) != null) {
+        networkRequestBundle.getRequestsToDrop().add(unqueuedRequest);
+      }
+      requestToServe = networkRequestBundle.getRequestToServe();
+      requestsToDrop = networkRequestBundle.getRequestsToDrop();
+
+      // Consume the first 8 bytes of requests since TCP impl uses this size to allocate buffer.
+      if (requestToServe != null) {
+        http2ServerMetrics.requestQueuingTime.update(System.currentTimeMillis() - requestToServe.getStartTimeInMs());
+        if (!consumeStream(requestToServe)) {
+          requestToServe = null;
+        }
+      }
+      Set<NetworkRequest> errorRequests = new HashSet<>();
+      for (NetworkRequest requestToDrop : requestsToDrop) {
+        http2ServerMetrics.requestQueuingTime.update(System.currentTimeMillis() - requestToDrop.getStartTimeInMs());
+        if (!consumeStream(requestToDrop)) {
+          errorRequests.add(requestToDrop);
+          logger.error("Error reading the stream, dropping the request: {}", requestToDrop);
+        }
+      }
+      requestsToDrop.removeAll(errorRequests);
+    } while (requestToServe == null && requestsToDrop.isEmpty());
+
+    return new NetworkRequestBundle(requestToServe, requestsToDrop);
+  }
+
   @Override
-  public NetworkRequest receiveRequest() throws InterruptedException {
-
-    NetworkRequest request = null;
-    while (request == null) {
-      request = requestQueue.take();
-      http2ServerMetrics.requestQueuingTime.update(System.currentTimeMillis() - request.getStartTimeInMs());
-      if (request.equals(EmptyRequest.getInstance())) {
-        logger.debug("Request handler {} received shut down command ", request);
-      } else {
-        DataInputStream stream = new DataInputStream(request.getInputStream());
-        try {
-          // The first 8 bytes is size of the request. TCP implementation uses this size to allocate buffer. See {@link BoundedReceive}
-          // Here we just need to consume it.
-          stream.readLong();
-        } catch (IOException e) {
-          logger.error("Encountered an error while reading length out, close the connection", e);
-          closeConnection(request);
-          request = null;
+  public List<NetworkRequest> getUnqueuedRequests() throws InterruptedException {
+    List<NetworkRequest> requestsToDrop = new ArrayList<>();
+    NetworkRequest unqueuedRequest;
+    do {
+      while ((unqueuedRequest = unqueuedRequests.poll()) != null) {
```

Comment:
hope the poll is blocking else this would be an infinite spinning loop if the drop queue is always empty..